### ISSUE TITLE
Observations on x-prediction vs v-prediction in Flow Matching (MNIST Trajectories)

### ISSUE THREAD

**Issue by Uljibuh** [LTH14/JiT#64]
First, I want to thank the authors for sharing their code and for the incredibly insightful "Back to Basics" paper. The shift toward $x$-prediction is a refreshing perspective on how we think about denoising and generative objectives. Great work!

I’ve been experimenting with the $x$-prediction objective proposed in the JiT paper but applied to Conditional Flow Matching (CFM) instead of standard Diffusion. While the paper focuses on $x$-prediction for Transformers in high-dimensional diffusion, I tested a lightweight U-Net on MNIST to observe the transport dynamics between source and target images ($x_0 \to x_1$).

<img width="1353" height="890" alt="Image" src="https://github.com/user-attachments/assets/26c33478-864e-43b7-bec5-065bfe78ec48" />

Key Observations:

Objective Compatibility: $x$-prediction performs remarkably well for Flow Matching in pixel space. Even with a small training budget (500 epochs), the model successfully learns to reconstruct the target image from a specific source image.

Architecture Sensitivity: I found that simple, shallow CNNs struggle significantly with both $x$ and $v$ prediction in this context. The hierarchical structure and skip connections of the U-Net seem essential for maintaining the coherence of the transport path and preserving digit identity.

Trajectory Bending (The "Elbow" Effect): A distinct behavior was observed in $x$-prediction: trajectories tend to bend significantly toward the end of the ODE integration ($t \to 1$).This is mathematically consistent with the velocity derivation: $v = \frac{\hat{x}_1 - x_t}{1-t}$.Even with numerical clamping (min=1e-3), the "bending" persists. This suggests that $x$-prediction inherently prioritizes endpoint accuracy/supervision at the cost of trajectory straightness.

Stability vs. Accuracy: While $v$-prediction remains the more "straight-path" objective (aligned with the principles of Rectified Flow), $x$-prediction feels more robust for manifold alignment in early training stages.

Implementation:I have modularized a minimal PyTorch implementation (data, models, sampling) to verify and visualize these dynamics:
[https://github.com/Uljibuh/velocity-vs-endpoint-flow-matching](url)
